content/docs/specs/native_histograms.md

```diff
@@ -0,0 +1,2353 @@
+  current and the previous sample.
+
+The following events trigger cutting a new chunk (for the reasons described in
+parentheses):
+
+- A change of sample type between integer histogram and float histogram
+  (because both require different chunk encodings altogether).
+- A change of sample type between gauge histogram and counter histogram
+  (because the leading byte has to denote the different type).
+- A counter reset for a counter histogram (to be stored in the leading byte as
+  counter reset information, see details below).
+- A schema change (which means we need a new chunk layout, and one and the same
+  chunk can only have one chunk layout).
+- A change of the zero threshold (which changes the chunk layout, see above).
+- A change of the custom values (which changes the chunk layout, see above).
+- A staleness marker is followed by a regular sample (which does not strictly
+  require a new chunk, but it can be assumed that most histograms will change
+  so much when they go away and come back that cutting a new chunk is the best
+  option).
+- The chunk size limit is exceeded (see [details below](#chunk-size-limit)).
+
+Differences in the spans would also change the chunk layout, but they are
+reconciled by adding (explicitly represented) unpopulated buckets as needed so
+that all histograms in a chunk share the same span structure. This is
+straightforward if a bucket disappears, because the missing bucket is simply
+added to the new histogram as an unpopulated bucket while the histogram is
+appended to the chunk. However, disappearance of a formerly populated bucket
+constitutes a counter reset (see [below](#counter-reset-considerations)), so
+this case can only happen for gauge histograms (which do not feature counter
+resets). The far more common case is that buckets exist in a newly appended
+histogram that did not exist in the previously appended histograms. In this
+case, these buckets have to be added as explicitly unpopulated buckets to all
+previously appended histograms. This requires a complete re-encoding of the
+entire chunk. (There is some optimization potential in only re-encoding the
+affected parts. Implementing this would be quite complicated. So far, the
+performance impact of the full re-encoding did not stick out as problematic.)
+
+### Staleness markers
+
+Staleness markers used to be simply a special float value among the many that
+can be used to represent the `NaN` value. The same float value could be used to
+mark a histogram time series as stale. However, this would require cutting a
+new chunk, just for the purpose of marking the series as stale, because a float
+value following a histogram value has to be stored in a different chunk (see
+above). Therefore, there is also a histogram version of a stale marker where
+the field for the sum of observations is set to the special stale `NaN` value.
+In this case, all other fields are ignored, so they can be set to values
```

Comment:
Having NaN observations could also result in the sum of observations being NaN according to
https://github.com/prometheus/docs/blob/2d3624ced474a0eeb70ec9a63d1d4ef7c96283a3/content/docs/specs/native_histograms.md?plain=1#L492-L496

How are the two cases (staleness markers vs NaN observations) distinguished? For the NaN observations case we wouldn't want to ignore the other fields.

Reply:
I would guess count==0 is the distinguishing feature, so not all fields are ignored

Reply:
There are lots of NaN values; we pick [one of them](https://github.com/prometheus/prometheus/blob/cd1f8ac129a289be8e7d98b6de57a9ba5814c406/model/value/value.go#L24-L28) to mean "stale".
It is not expected that the same value can be generated by accident.

Reply:
The count value shouldn't matter. If the sum is the specific bit pattern that defines a staleness marker, the whole histogram must be treated as a staleness marker.

Reply:
This is super confusing :( Not sure how we can improve the text, but it seems like having `NaN` written here is the source of trouble. Let me suggest something in a new comment to try and clarify.

Reply:
I'll resolve this conversation, because we have a new one below with @krajorama 's suggestion.